### PR TITLE
Add Multi Block Change API

### PR DESCRIPTION
**Pull request by brodybeckwith**
API to send large amounts of fake block changes in one packet per chunk section.

### CODE REVIEW

File: patches/server/0852-Multi-Block-Change-API-Implementation.patch [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
++
++            Map<Short, BlockState> sectionData = sectionMap.computeIfAbsent(sectionPos, key -> new HashMap<>());
++            sectionData.put(SectionPos.sectionRelativePos(blockPos), ((CraftBlockData) blockData).getState());
++        }
++
++        for (Map.Entry<SectionPos, Map<Short, BlockState>> entry : sectionMap.entrySet()) {
++            SectionPos sectionPos = entry.getKey();
++            Map<Short, BlockState> blockData = entry.getValue();
++
++            FriendlyByteBuf friendlyByteBuf = new FriendlyByteBuf(Unpooled.buffer());
```
Comment (kennytv) [inline patches/server/0852-Multi-Block-Change-API-Implementation.patch]: Manual writing is a nono here - either make a new ClientboundSectionBlocksUpdatePacket constructor or temporarily modify the `states` block container in LevelChunkSection

File: patches/api/0357-Multi-Block-Change-API.patch [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ 
++    // Paper start
++    /**
++     * Send multiple block changes. This fakes a multi block change packet for each
++     * chunk section that a block change occurs. This will not actually change the world in any way.
++     *
++     * @param blockChanges A map of the locations you want to change to their new block data
++     * @param suppressLightUpdates Whether to suppress light updates or not
++     */
++    public void sendMultiBlockChange(@NotNull Map<Location, BlockData> blockChanges, boolean suppressLightUpdates);
```
Comment (kennytv) [inline patches/api/0357-Multi-Block-Change-API.patch]: Another default method calling `sendMultiBlockChange(blockChanges, false)` would be nice

**Review by kennytv — CHANGES_REQUESTED** [review]
Use fully qualified imports please (at least for the ones used less than like 5 times), to reduce conflicts on updates

File: patches/server/0852-Multi-Block-Change-API-Implementation.patch (hunk 1 above)
Comment (brodybeckwith) [reply patches/server/0852-Multi-Block-Change-API-Implementation.patch]: I went ahead and made a new constructor in latest changes, wasn't sure if I should try and use what was available to do it or just make a new constructor initially.

File: patches/api/0357-Multi-Block-Change-API.patch (hunk 2 above)
Comment (brodybeckwith) [reply patches/api/0357-Multi-Block-Change-API.patch]: Added.

**Comment by brodybeckwith** [comment]
> Use fully qualified imports please (at least for the ones used less than like 5 times), to reduce conflicts on updates

Should be good now, I added the things you suggested and fixed imports I believe.

File: patches/server/0852-Multi-Block-Change-API-Implementation.patch [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
++
++        for (Map.Entry<Location, BlockData> entry : blockChanges.entrySet()) {
++            Location location = entry.getKey();
++            if (!location.getWorld().equals(this.getWorld())) continue;
++
++            BlockData blockData = entry.getValue();
++            BlockPos blockPos = new BlockPos(location.getBlockX(), location.getBlockY(), location.getBlockZ());
++            SectionPos sectionPos = SectionPos.of(blockPos);
++
++            Map<Short, net.minecraft.world.level.block.state.BlockState> sectionData = sectionMap.computeIfAbsent(sectionPos, key -> new it.unimi.dsi.fastutil.shorts.Short2ObjectArrayMap<>());
```
Comment (kennytv) [inline patches/server/0852-Multi-Block-Change-API-Implementation.patch]:
```suggestion
+            it.unimi.dsi.fastutil.shorts.Short2ObjectMap<net.minecraft.world.level.block.state.BlockState> sectionData = sectionMap.computeIfAbsent(sectionPos, key -> new it.unimi.dsi.fastutil.shorts.Short2ObjectArrayMap<>());
```

File: patches/server/0852-Multi-Block-Change-API-Implementation.patch [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+index 82ea4fabd5732052a286d50bcff8bbcc2c4aa7d7..53aa2082f1b99f841c2ede6c6853b29d56bab458 100644
+--- a/src/main/java/net/minecraft/network/protocol/game/ClientboundSectionBlocksUpdatePacket.java
++++ b/src/main/java/net/minecraft/network/protocol/game/ClientboundSectionBlocksUpdatePacket.java
+@@ -54,6 +54,16 @@ public class ClientboundSectionBlocksUpdatePacket implements Packet<ClientGamePa
+ 
+     }
+ 
++    // Paper start
++    public ClientboundSectionBlocksUpdatePacket(SectionPos sectionPos, it.unimi.dsi.fastutil.shorts.Short2ObjectMap<BlockState> blockChanges, boolean suppressLightUpdates)
++    {
```
Comment (kennytv) [inline patches/server/0852-Multi-Block-Change-API-Implementation.patch]: code style

**Review by kennytv — APPROVED** [review]
Other than those two small things, looks good

File: patches/server/0852-Multi-Block-Change-API-Implementation.patch (hunk 4 above)
Comment (brodybeckwith) [reply patches/server/0852-Multi-Block-Change-API-Implementation.patch]: Oops Allman style :)

**Comment by brodybeckwith** [comment]
> Other than those two small things, looks good

Should be all fixed now.

**Pull request merged**